File: src/main/java/de/pbauerochse/worklogviewer/youtrack/issuedetails/IssueField.java
package de.pbauerochse.worklogviewer.youtrack.issuedetails;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

/**
 * Created by patrick on 31.10.15.
 * "field":[{"name":"resolved","value":"1446292986266"}]
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public class IssueField {

    private String name;
    private String value;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getValue() {
        return value;
    }

    public void setValue(String value) {
        this.value = value;
    }
}

File: connectors/v20182/src/main/java/de/pbauerochse/worklogviewer/connector/v2018/domain/issue/IssueField.kt
package de.pbauerochse.worklogviewer.connector.v2018.domain.issue

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ArrayNode
import com.fasterxml.jackson.databind.node.ObjectNode

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
class IssueField @JsonCreator constructor(
    @JsonProperty("name") val name: String,
    @JsonProperty("value") val value: JsonNode
) {

    val textValue: String? by lazy {
        if (value.isTextual) value.textValue() else null
    }

    fun toGroupByKey(): String? = toGroupByKey(value)

    private fun toGroupByKey(node: JsonNode): String? = when {
        node.isTextual -> node.textValue()
        node.isArray -> toGroupByKey(node as ArrayNode)
        node.isObject -> toGroupByKey(node as ObjectNode)
        else -> throw IllegalStateException("Unknown node type ${value.nodeType} for value $value")
    }

    private fun toGroupByKey(node: ArrayNode): String? = node
        .map { toGroupByKey(it) }
        .filter { it?.isNotBlank() ?: false }
        .joinToString { "," }

    private fun toGroupByKey(node: ObjectNode): String? {
        return when {
            node.has("value") -> toGroupByKey(node.get("value"))
            else -> null
        }
    }

}